light_node/src/file_config.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2015 Andrew Gallant
+// Copyright (c) 2019 SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+
+// This module provides routines for reading ripgrep config "rc" files. The
+// primary output of these routines is a sequence of arguments, where each
+// argument corresponds precisely to one shell argument.
+
+use std::error::Error;
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+
+use Result;
+
+macro_rules! errored {
+  ($($tt:tt)*) => {
+      return Err(From::from(format!($($tt)*)));
+  }
+}
+
+/// Return a sequence of arguments derived from ripgrep rc configuration files.
+///
+/// If no_messages is false and there was a problem reading a config file,
+/// then errors are printed to stderr.
+pub fn args(config_path: PathBuf, no_messages: bool) -> Vec<OsString> {
+    let (args, errs) = match parse(&config_path) {
+        Ok((args, errs)) => (args, errs),
+        Err(err) => {
+            if !no_messages {
+                eprintln!("{}", err);
+            }
+            return vec![];
+        }
+    };
+    if !no_messages && !errs.is_empty() {
+        for err in errs {
+            eprintln!("{}:{}", config_path.display(), err);
+        }
+    }
+    // println!("{}: arguments loaded from config file: {:?}", config_path.display(), args);
+    
+    args
+}
+
+/// Parse a single ripgrep rc file from the given path.
+///
+/// On success, this returns a set of shell arguments, in order, that should
+/// be pre-pended to the arguments given to ripgrep at the command line.
+///
+/// If the file could not be read, then an error is returned. If there was
+/// a problem parsing one or more lines in the file, then errors are returned
+/// for each line in addition to successfully parsed arguments.
+fn parse<P: AsRef<Path>>(
+    path: P,
+) -> Result<(Vec<OsString>, Vec<Box<dyn Error>>), Box<dyn Error>> {
+    let path = path.as_ref();
+    match File::open(&path) {
+        Ok(file) => parse_reader(file),
+        Err(err) => errored!("{}: {}", path.display(), err),
+    }
+}
+
+/// Parse a single ripgrep rc file from the given reader.
+///
+/// Callers should not provided a buffered reader, as this routine will use its
+/// own buffer internally.
+///
+/// On success, this returns a set of shell arguments, in order, that should
+/// be pre-pended to the arguments given to ripgrep at the command line.
+///
+/// If the reader could not be read, then an error is returned. If there was a
+/// problem parsing one or more lines, then errors are returned for each line
+/// in addition to successfully parsed arguments.
+fn parse_reader<R: io::Read>(
+    rdr: R,
+) -> Result<(Vec<OsString>, Vec<Box<dyn Error>>), Box<dyn Error>> {
```

Comment:
For code reader it's hard to track what tuples returned from functions are doing.
Instead of returning tuple you could do:
```rust
fn parse_reader<R: io::Read>(
    rdr: R,
) -> Result<Vec<Result<String, ParseError>>,  ReaderError>
```